### PR TITLE
add support for Japanese (JA)

### PR DESCRIPTION
**Pull request by siikamiika**
### Changes proposed in this pull request:

* Add support for JA

### Status

- [x] READY
- [ ] HOLD
- [ ] WIP (Work-In-Progress)

### How to verify this change

```python
import num2words
num2words.num2words(123456, lang='ja')
num2words.num2words(123456, lang='ja', reading=True)
num2words.num2words(-5, lang='ja')
num2words.num2words(0, lang='ja', reading=True)
num2words.num2words(74, lang='ja', reading=True)
num2words.num2words(74, lang='ja', reading=True, prefer=['し', 'しち'])
num2words.num2words(123456, lang='ja', reading=True, to="ordinal")
num2words.num2words(123456, lang='ja', reading=True, to="ordinal_num")
```

### Additional notes

-

### CODE REVIEW

**Comment by erozqba** [comment]
Hi @siikamiika 
Thanks a lot for taking the time to make this PR! It's really appreciated.
I don't know any Japanese so we really trust the contributors that make the PR.
I can see this is a work in progress, as a suggestion, don't forget to update the README.rst and to implement as many automatic tests as you can.

Thanks a lot again!

**Comment by ventilooo** [comment]
hey, @erozqba ! I speak a bit of Japanese.
I'll try to handle the review as much as I can :wink: 

@siikamiika ping me once the WIP is over !
Thanks again for your contribution :+1:

**Comment by siikamiika** [comment]
Seems like the code is having some issues with Python 2. Probably Unicode related. The tests should be fine for Python 3 now, but I haven't updated README yet.

**Comment by siikamiika** [comment]
@erozqba @ventilooo Everything looks fine to me now. This sure was an interesting language to implement. :sweat_smile: I'm also just learning Japanese so I did this as an exercise. I did some research on the preferred readings of some number words in Japanese, but there could be errors somewhere.

File: num2words/lang_JA.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+    (686, ("朱鳥", "しゅちょう")),
+    (701, ("大宝", "たいほう")),
+    (704, ("慶雲", "けいうん")),
+    (708, ("和銅", "わどう")),
+    (715, ("霊亀", "れいき")),
+    (717, ("養老", "ようろう")),
+    (724, ("神亀", "じんき")),
+    (729, ("天平", "てんぴょう")),
+    # (749, ("天平感宝", "てんぴょうかんぽう")),
+    (749, ("天平勝宝", "てんぴょうしょうほう")),
```
Comment (erozqba) [inline num2words/lang_JA.py]: Why there are some commented lines as this one in this list?

**Comment by siikamiika** [comment]
It's because there have been multiple emperors/empresses during the same
year and one has to be chosen since num2words doesn't distinguish between
months and days. The last one is the one who has reigned until the next
emperor/empress. I actually had them the wrong way around first (used the
first one) but fixed them in a commit

**Comment by siikamiika** [comment]
@ventilooo 
> Instead of putting some emperor/empress in comment, could you solve this with a method ?
> 
> Like with :arrow_down:
> 
> self.assertEqual(n2j(0, prefer=["〇"]), "〇")
> 
> In order to easily give the choice to the user.

Sounds like a good idea, I'll try to do that.

> Also I don't know how useful this could be, but having the option to choose the the output format between
> 
>  *   Romaji
>  *   Kanji
>  *   Hiragana
> 
> If I want to write an app for kid, they don't necessarily know all the kanji associated with numbers.
> Or if I want to write an app for foreigner, to learn Japanese. Having an romaji output would help.

Hiragana output is currently supported with the `reading` flag, but I'm not sure if that's the best way to choose it if you want to choose output between multiple different writing systems. There's also katakana, which would be pretty easy to implement if I remember the Unicode code points correctly.

Romaji is a bit complicated, though. First, there are multiple ways to romanize Japanese text, as seen from your link. An optional dependency to something like [python-romkan](https://github.com/soimort/python-romkan) could be added, but even then there would be the problem of separating the number words with spaces only for romanized Japanese but not with kana. Anyway, I agree that there could be some use for that feature. I guess I could look into implementing romaji at some point.

**Comment by siikamiika** [comment]
I just realized that the era thing wasn't as simple as having some ruler for less than a year. The `prefer` suddenly makes a lot more sense but I might replace the year in the tuples with another tuple containing both the start and the end of the era.

Some reading:
https://en.wikipedia.org/wiki/Eiwa
https://en.wikipedia.org/wiki/Tenju

**Pull request merged**